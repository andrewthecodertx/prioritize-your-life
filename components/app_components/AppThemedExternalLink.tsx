import { Href, Link } from "expo-router";
import { Platform } from "react-native";
import { openBrowserAsync } from "expo-web-browser";
import { type ComponentProps } from "react";
import { useThemeColor } from "@/hooks/useThemeColor";
import { COLORTHEME } from "@/constants/Colors";

type Props = Omit<ComponentProps<typeof Link>, "href"> & {
  href: Href<string | object>;
};

/**
 * Renders a themed external link component.
 * @param {Props} props - The component props.
 * @param {string} props.href - The URL of the external link.
 * @param {React.CSSProperties} props.style - The custom style for the link.
 * @param {React.HTMLAttributes<HTMLAnchorElement>} props.rest - Additional HTML attributes for the link.
 * @returns {JSX.Element} The rendered themed external link component.
 */
export function AppThemedExternalLink({ href, style, ...rest }: Props) {
  const color = useThemeColor(
    { light: COLORTHEME.light.background, dark: COLORTHEME.dark.background },
    "background"
  );
  return (
    <Link
      target="_blank"
      {...rest}
      href={href}
      onPress={async (event) => {
        if (Platform.OS !== "web") {
          // Prevent the default behavior of linking to the default browser on native.
          event.preventDefault();
          // Open the link in an in-app browser.
          await openBrowserAsync((href as string));
        }
      }}
      style={[{ color }, style]}
    />
  );
}
