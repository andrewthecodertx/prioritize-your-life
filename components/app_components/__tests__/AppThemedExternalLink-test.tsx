import React from "react";
import { Platform } from "react-native";
import { render, fireEvent } from "@testing-library/react-native";
import renderer, { act } from "react-test-renderer";
import { AppThemedExternalLink } from "../AppThemedExternalLink";
import { openBrowserAsync } from "expo-web-browser";


describe("AppThemedExternalLink", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <AppThemedExternalLink href="https://www.google.com">
          Snapshot test!
        </AppThemedExternalLink>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("onPress navigates to another page on web platforms", async () => {
    Platform.OS = "web"; // Simulate a web platform

    const href = "https://www.google.com";

    const { getByText } = render(
      <AppThemedExternalLink href={href}>Test Link</AppThemedExternalLink>
    );

    const link = getByText("Test Link");
    fireEvent.press(link);

    expect(openBrowserAsync).not.toHaveBeenCalled();
  });
  it("onPress navigates to an external link on native platforms", async () => {
    Platform.OS = "ios"; // Simulate a native platform

    const href = "https://www.google.com";
    const mockEvent = {
      preventDefault: jest.fn(),
    };

    const { getByText } = render(
      <AppThemedExternalLink href={href}>Test Link</AppThemedExternalLink>
    );

    const link = getByText("Test Link");
    fireEvent.press(link, mockEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalled();
    expect(openBrowserAsync).toHaveBeenCalledWith(href);
  });
});
