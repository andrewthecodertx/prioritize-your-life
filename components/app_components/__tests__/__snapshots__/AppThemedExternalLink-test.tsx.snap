// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemedExternalLink renders correctly 1`] = `
<Text
  href="https://www.google.com"
  onPress={[Function]}
  role="link"
  style={
    [
      {
        "color": "#000000",
      },
      undefined,
    ]
  }
>
  Snapshot test!
</Text>
`;
